zookeeper/core/component.py

```diff
@@ -303,6 +303,106 @@ def wrapped_fn(instance) -> List[str]:
     component_cls.__dir__ = wrapped_fn
 
 
+######################################
+# Implement the `ItemsView` protocol #
+######################################
+
+# Implementing these three methods makes `dict(component_instance)`,
+# `list(component_instance)`, et cetera be something sensible.
+#
+# In particular, `dict(component_instance)` has field names as keys and field
+# values as values, giving a flattened view of the entire component tree
+# including all sub-components. This dictionary is precisely the configuration
+# that is required to completely re-create the component, including all
+# sub-components, i.e. `configure(new_instance, dict(old_instance))`.
+
+
+def _wrap_len(component_cls: Type) -> None:
+    if hasattr(component_cls, "__len__"):
+        raise TypeError("Component classes must not define a custom `__len__` method.")
```

Comment:
*Technically* this is a breaking change, but I strongly doubt that anybody was using component classes with custom `__len__`, `__contains__`, or `__iter__` methods.

I'm happy to make a new minor version release but I think we can be slightly flexible about semver here and not worry about a new major version.